src/types.ts

```diff
@@ -1,38 +1,31 @@
 export type UnbindFn = () => void;
 
-type AnyFunction = (...args: any[]) => any;
-
-type GetEventType<Target extends EventTarget, Type extends string> = Target extends unknown
-  ? `on${Type}` extends keyof Target
-    ? GetEventTypeFromListener<
-        // remove types that aren't assignable to `AnyFunction`
-        // so that we don't end up with union like `MouseEvent | Event`
-        Extract<Target[`on${Type}`], AnyFunction>
-      >
+type ExtractEventTypeFromHandler<MaybeFn extends unknown> = MaybeFn extends (
+  this: any,
+  event: infer MaybeEvent,
+) => any
+  ? MaybeEvent extends Event
+    ? MaybeEvent
     : Event
   : never;
 
-type GetEventTypeFromListener<T extends AnyFunction> = T extends (this: any, event: infer U) => any
-  ? U extends Event
-    ? U
-    : Event
-  : Event;
+type GetEventType<Target extends EventTarget, EventName extends string> =
+  // Does the target have a on${eventName} property?
+  `on${EventName}` extends keyof Target
+    ? ExtractEventTypeFromHandler<Target[`on${EventName}`]>
+    : Event;
 
-export type Binding<Target extends EventTarget = EventTarget, Type extends string = string> = {
-  type: Type;
-  listener: Listener<GetEventType<Target, Type>, Target>;
+export type Binding<Target extends EventTarget = EventTarget, EventName extends string = string> = {
+  type: EventName;
+  listener: Listener<Target, EventName>;
   options?: boolean | AddEventListenerOptions;
 };
 
-export type Listener<Ev extends Event, Target extends EventTarget> =
-  | ListenerObject<Ev>
-  // For a listener function, the `this` binding is the target the event listener is added to
-  // using bivariance hack here so if the user
-  // wants to narrow event type by hand TS
-  // won't give them an error
-  | { bivarianceHack(this: Target, e: Ev): void }['bivarianceHack'];
```

Comment:
I got rid of this `bivarianceHack`. I could not create a test case where this was helpful

Reply:
@Ayub-Begimkulov can you speak to this one?

Reply:
Hey, Alex. Just noticed an email from GitHub.

The `bivarianceHack` is useful in cases where event target doesn't have a `on...` property for the event that we attach listener to.

For instance:

```
bind(element, {
  type: 'asdf',
  listener(event) {
    type EventType = typeof event; // Event
  }
})
```

As you can see, in such cases we fallback to a default `Event` type, which is correct, however, the problem is that we can't explicitly narrow this type, if we want.

Here is an example:

```ts
bind(element, {
  type: 'asdf',
  listener(event: MouseEvent) {
  // ^^^^^^^^^^^^^^^^^^^^
  // Type '(e: MouseEvent) => void' is not assignable to type '(e: Event) => void'
  }
})
```

This type error makes sense if you think about it, but it doesn't allow us to have a polymorphism, that's why TS team made an exception for the objects. And it's was a reason why I made such strange type definition.

I agree that it's a rear case, but since it's a library, I think that it might be useful to cover all the cases, especially if it's not that difficult.

Reply:
My bad, just tried a new version, and it seems that everything works without the hack. 

I guess it's because the second argument to the `option` function is an object itself.

Haven't thought about it initially, haha. If it works without it - it's even better!

Nice job in noticing it!

Reply:
Thanks so much for having a good look!!!!